### ISSUE TITLE
macos desktop Not able to display textures using FlutterTextureRegistry anymore on latest master, but works on stable.

### ISSUE THREAD

**Issue by larryaasen** [flutter/flutter#81120]
After upgrading Flutter from stable version 2.0.5 to master, 2.2.0-11.0.pre.260, my macOS desktop app was not able to display textures using FlutterTextureRegistry anymore. It started displaying errors. For this breaking change with textures, is there a PR or document with details on upgrading?
```
2021-04-23 18:11:33.948993-0400 dive_app[20133:3434035] Could not create Metal texture from pixel buffer: CVReturn -6660
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177215456 did not return a valid texture.
2021-04-23 18:11:33.949215-0400 dive_app[20133:3434035] Could not create Metal texture from pixel buffer: CVReturn -6660
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177218048 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177456688 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177471152 did not return a valid texture.
2021-04-23 18:11:33.949575-0400 dive_app[20133:3434035] Could not create Metal texture from pixel buffer: CVReturn -6660
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177426416 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177215456 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177218048 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177456688 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177471152 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177426416 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177215456 did not return a valid texture.
2021-04-23 18:11:33.982847-0400 dive_app[20133:3434035] Could not create Metal texture from pixel buffer: CVReturn -6660
...
```

<details>
  <summary>Logs</summary>

```
2021-04-23 18:11:33.948993-0400 dive_app[20133:3434035] Could not create Metal texture from pixel buffer: CVReturn -6660
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177215456 did not return a valid texture.
2021-04-23 18:11:33.949215-0400 dive_app[20133:3434035] Could not create Metal texture from pixel buffer: CVReturn -6660
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177218048 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177456688 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177471152 did not return a valid texture.
2021-04-23 18:11:33.949575-0400 dive_app[20133:3434035] Could not create Metal texture from pixel buffer: CVReturn -6660
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177426416 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177215456 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177218048 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177456688 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177471152 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177426416 did not return a valid texture.
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 105553177215456 did not return a valid texture.
2021-04-23 18:11:33.982847-0400 dive_app[20133:3434035] Could not create Metal texture from pixel buffer: CVReturn -6660
```

```
Analyzing dive_app...                                                   
No issues found! (ran in 4.0s)
```

```
[✓] Flutter (Channel master, 2.2.0-11.0.pre.260, on macOS 11.2.3 20D91 darwin-x64, locale en-US)
    • Flutter version 2.2.0-11.0.pre.260 at /Applications/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 59fc92168a (15 hours ago), 2021-04-23 22:47:40 +0200
    • Engine revision 6fa0fb0059
    • Dart version 2.14.0 (build 2.14.0-18.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/larry/Library/Android/sdk
    • Platform android-29, build-tools 28.0.3
    • ANDROID_HOME = /Users/larry/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.2.3 20D91 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 90.0.4430.85
    ! Error: Apple TV is not connected. Xcode will continue when Apple TV is connected. (code -13)

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
@larryaasen 
Can you provide a minimal reproducible code sample that shows this issue ?
Thanks.

**Comment by larryaasen** [comment]
@darshankawar I have a very large project that utilizes ffi and a large C library. I removed that library from the app, cut out lots of code, and created a more simplified demo. I'm still wrapping that up and will try to get you some code soon.

**Comment by larryaasen** [comment]
@darshankawar  I've uploaded a [demo of this problem here](https://drive.google.com/file/d/1ZdCAPiLRkJZ5m2SwgWgnmn0fJjpwcc7w/view?usp=sharing). Also, here are two screenshots, one on stable and one on master, where you can see the issue:
```
Could not create Metal texture from pixel buffer: CVReturn -6660
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 140441962111248 did not return a valid texture.
```
<img width="1308" alt="Screen Shot 2021-05-05 at 8 10 35 AM" src="https://user-images.githubusercontent.com/325973/117295372-c1a73480-ae41-11eb-956f-11a71db1f771.png">
<img width="1354" alt="Screen Shot 2021-05-06 at 8 00 01 AM" src="https://user-images.githubusercontent.com/325973/117295389-c7047f00-ae41-11eb-9cc2-20581f5d93f6.png">

The original app is large and uses ffi with a large C library that I removed from the demo. I captured one frame of video for this demo to make it easier to debug.

**Comment by darshankawar** [comment]
@larryaasen 
Thanks for taking effort in creating a minimal repro code sample. I ran it and do see the mentioned logs in the console after running the app.


```
call.method= initializeTexture
addFrameCapture: added texture source: fa8edab0-ae68-11eb-a376-8f312d983990
call.method= createVideoMix
flutter: videoMix=DivePreview
2021-05-06 18:15:47.079 dive_ui_example[25651:380961] Could not create Metal texture from pixel buffer: CVReturn -6660
[ERROR:flutter/shell/platform/embedder/embedder_external_texture_metal.mm(59)] External texture callback for ID 140482139775392 did not return a valid texture.
The Flutter DevTools debugger and profiler on macOS is available at: http://127.0.0.1:9100?uri=http%3A%2F%2F127.0.0.1%3A58037%2FYohAsY2CBug%3D%2F

```

I just have one more question for you. I see that you are using `flutter_riverpod` [plugin](https://pub.dev/packages/flutter_riverpod/install) in the example app, which is 3rd party. Is it possible to remove the plugin related code and check if issue still persists without it ?  With it, the `AppWidget` class seems to be wrapped with `ProviderScope` which is coming from the plugin.

**Comment by larryaasen** [comment]
@darshankawar I removed Riverpod from the demo app and uploaded a new copy of the project here.
https://drive.google.com/file/d/1uksanVXAcgCXwhmXOZWZ8BWOqlfnNTnD/view?usp=sharing

**Comment by darshankawar** [comment]
Thanks for the updated code sample. Verified this on latest master and throws below log. Also it doesn't seem to render the image on screen as shown below:

1. Master:

<img width="1188" alt="Screenshot 2021-05-07 at 5 39 12 PM" src="https://user-images.githubusercontent.com/67046386/117448149-feffe500-af5b-11eb-8906-f93a761e61c3.png">

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.3.0-1.0.pre.242, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.3.0-1.0.pre.242 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3ab799bbc9 (2 hours ago), 2021-05-06 19:49:02 -0700
    • Engine revision f57e986aa8
    • Dart version 2.14.0 (build 2.14.0-74.0.dev)

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • Mac OS X 10.15.4 19E2269
      darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 90.0.4430.93

! Doctor found issues in 1 category.
dhs@Dhss-MacBook-Pro ~ % 

```

</details>

The same code works on latest stable and renders on screen with no log messages, as below:

<img width="1111" alt="Screenshot 2021-05-07 at 5 35 18 PM" src="https://user-images.githubusercontent.com/67046386/117448291-2fe01a00-af5c-11eb-9f5e-c07f623872bb.png">


<details>
<summary>flutter doctor -v</summary>

```

[✓] Flutter (Channel stable, 2.0.6, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.6 at /Users/dhs/documents/fluttersdk/flutter
    • Framework revision 1d9032c7e1 (7 days ago), 2021-04-29 17:37:58 -0700
    • Engine revision 05e680e202
    • Dart version 2.12.3

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 90.0.4430.93

! Doctor found issues in 1 category.

```

</details>

**Comment by lsegal** [comment]
Could this be bumped? Looks like due to the slow triage this regression snuck its way into the latest stable release, meaning unless I am missing something, [Texture](https://api.flutter.dev/flutter/widgets/Texture-class.html) is entirely broken on macOS right now.

See https://github.com/flutter/plugins/pull/3712#issuecomment-828295684

Looking through 2.2.0 release notes, is this possibly related to https://github.com/flutter/engine/pull/24157? Seems like that feature involved at least one revert and re-revert already.

**Comment by cbenhagen** [comment]
@lsegal I wouldn't say it is completely broken. It has a huge [memory leak](https://github.com/flutter/flutter/issues/79648) in stable and the corresponding [cherry pick request](https://github.com/flutter/flutter/issues/81308) unfortunately never made it into the stable or beta channel.

If you ignore the memory leak and the errors in the log, the [macOS video player](https://github.com/flutter/plugins/pull/3712) plugin which is using textures works for me on all channels. It looks like the logged errors could in fact just be ignored as they are being handled silently in the GL version. Let me open a PR for that to start the discussion.

@larryaasen would you mind sharing your example as a GitHub repo? I can't access it to check why it is not working. You might want to check my video player implementation to see why your usage of textures does not work anymore. Maybe you need to add the `kCVPixelBufferMetalCompatibilityKey`?

**Comment by boyan01** [comment]
> Maybe you need to add the `kCVPixelBufferMetalCompatibilityKey`?

👍 amazing, I add the keys and then it worked. thansks @cbenhagen 

https://github.com/cbenhagen/plugins/blob/13e55fbef29f8239b5a3b3222634069b7153fc6f/packages/video_player/video_player_macos/macos/Classes/FLTVideoPlayerPlugin.m#L155-L159

**Comment by cbenhagen** [comment]
@boyan01 you're welcome!

I opened https://github.com/flutter/flutter/issues/84855 and https://github.com/flutter/engine/pull/26816 to discuss removal of the error messages.

**Comment by larryaasen** [comment]
> @larryaasen would you mind sharing your example as a GitHub repo? I can't access it to check why it is not working. You might want to check my video player implementation to see why your usage of textures does not work anymore. Maybe you need to add the `kCVPixelBufferMetalCompatibilityKey`?

I made the GitHub repo of this demo public here: https://github.com/larryaasen/dive_demo
Note that this is just a demo of the problem and is not a releasable app.

**Comment by larryaasen** [comment]
> Maybe you need to add the `kCVPixelBufferMetalCompatibilityKey`?
I tried that add it did not work.

**Comment by cbenhagen** [comment]
It might be that you can't use `CVPixelBufferCreateWithBytes` with `kCVPixelBufferIOSurfacePropertiesKey`. At least according to an old document this was once the case: https://developer.apple.com/library/archive/qa/qa1781/_index.html

**Comment by cbenhagen** [comment]
@larryaasen your main issue is this: `Could not create Metal texture from pixel buffer: CVReturn -6660`

**Comment by larryaasen** [comment]
@cbenhagen Since this works on stable versions 2.0.5 and 2.0.6, but not on master, 2.2.0-11.0.pre.260, there must have been a breaking change related to the metal updates for textures. Was there any upgrading details that were provided on how to make updates for this change? The addition of kCVPixelBufferMetalCompatibilityKey did not work. The repo https://github.com/larryaasen/dive_demo has an example of this issue.

**Comment by cbenhagen** [comment]
@larryaasen I would try to get rid of `CVPixelBufferCreateWithBytes` and add `kCVPixelBufferIOSurfacePropertiesKey` to the  attributes dictionary.

**Comment by gspencergoog** [comment]
@iskakaushik Do you have any information on this issue?

**Comment by iskakaushik** [comment]
I tried the https://github.com/larryaasen/dive_demo it seems to be working. I suspect that this was fixed by https://github.com/flutter/engine/pull/27506. @larryaasen mind giving it a quick try?

**Comment by larryaasen** [comment]
> I tried the https://github.com/larryaasen/dive_demo it seems to be working. I suspect that this was fixed by [flutter/engine#27506](https://github.com/flutter/engine/pull/27506). @larryaasen mind giving it a quick try?

I posted this on the demo README: Update (8/15/2021) - this demo now works on channel master 2.4.0-5.0.pre.181.

**Comment by iskakaushik** [comment]
Thanks! If you get a chance, please submit an issue for the new problem you are running into.
